src/AdoPat/PatCache.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AdoPat
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Text.Json;
+    using Microsoft.VisualStudio.Services.DelegatedAuthorization;
+
+    /// <summary>
+    /// A simple cache mapping <see cref="string"/> to <see cref="PatToken"/> backed by secure storage.
+    /// </summary>
+    public class PatCache
+    {
+        private IStorageWrapper storage;
+        private Lazy<Dictionary<string, PatToken>> cache;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PatCache"/> class.
+        /// </summary>
+        /// <param name="storage">A class which implements <see cref="IStorageWrapper"/>.</param>
+        public PatCache(IStorageWrapper storage)
+        {
+            // Note: This assumes any required persistance validity has been
+            // checked beforehand.
+            this.storage = storage;
+
+            // The cache is lazy so that the initial cache read from the
+            // underlying storage only happens once and subsequent reads
+            // operate on the in-memory value.

```

Comment:
This comment may be more suited to the first use of lazy, so a reader finds it above first.